homeassistant/components/lg_thinq/vacuum.py

```diff
@@ -0,0 +1,185 @@
+    """State of device."""
+
+    HOMING = "homing"
+    PAUSE = "pause"
+    RESUME = "resume"
+    SLEEP = "sleep"
+    START = "start"
+    WAKE_UP = "wake_up"
+
+
+ROBOT_STATUS_TO_HA = {
+    "charging": STATE_DOCKED,
+    "diagnosis": STATE_IDLE,
+    "homing": STATE_RETURNING,
+    "initializing": STATE_IDLE,
+    "macrosector": STATE_IDLE,
+    "monitoring_detecting": STATE_IDLE,
+    "monitoring_moving": STATE_IDLE,
+    "monitoring_positioning": STATE_IDLE,
+    "pause": STATE_PAUSED,
+    "reservation": STATE_IDLE,
+    "setdate": STATE_IDLE,
+    "sleep": STATE_IDLE,
+    "standby": STATE_IDLE,
+    "working": STATE_CLEANING,
+    "error": STATE_ERROR,
+}
+ROBOT_BATT_TO_HA = {
+    "moveless": 5,
+    "dock_level": 5,
+    "low": 30,
+    "mid": 50,
+    "high": 90,
+    "full": 100,
+    "over_charge": 100,
+}
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ThinqConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up an entry for vacuum platform."""
+    entities: list[ThinQStateVacuumEntity] = []
+    for coordinator in entry.runtime_data.coordinators.values():
+        if (
+            descriptions := DEVICE_TYPE_VACUUM_MAP.get(
+                coordinator.api.device.device_type
+            )
+        ) is not None:
+            for description in descriptions:
+                entities.extend(
+                    ThinQStateVacuumEntity(coordinator, description, property_id)
+                    for property_id in coordinator.api.get_active_idx(description.key)
+                )
+
+    if entities:
+        async_add_entities(entities)
+
+
+class ThinQStateVacuumEntity(ThinQEntity, StateVacuumEntity):
+    """Represent an thinq vacuum platform."""
+
+    def __init__(
+        self,
+        coordinator: DeviceDataUpdateCoordinator,
+        entity_description: StateVacuumEntityDescription,
+        property_id: str,
+    ) -> None:
+        """Initialize vacuum platform."""
+        super().__init__(coordinator, entity_description, property_id)
+
+        self._attr_supported_features = (
+            VacuumEntityFeature.SEND_COMMAND
+            | VacuumEntityFeature.STATE
+            | VacuumEntityFeature.BATTERY
+            | VacuumEntityFeature.START
+            | VacuumEntityFeature.PAUSE
+            | VacuumEntityFeature.RETURN_HOME
+        )
+        self._device_state: str | None = None
+
+    def _update_status(self) -> None:
+        """Update status itself."""
+        super()._update_status()
+
+        # Update state.
+        self._device_state = self.data.current_state
+        if self._device_state:
+            self._attr_state = ROBOT_STATUS_TO_HA.get(self._device_state)
+
+        # Update battery.
+        if (level := self.data.battery) is not None:
+            self._attr_battery_level = (
+                level if isinstance(level, int) else ROBOT_BATT_TO_HA.get(level, 0)
+            )
```

Comment:
Do we also have a separate sensor for this?

Reply:
This vaccum is only for ROBOT_CLEANER, we just have battery_sensor for STICK_CLEANER.

https://github.com/home-assistant/core/pull/125799/commits/36e7f3949b6734e71408023084561767a9133d31#diff-f5c6c1b249226c3c3a41b22061268769e68a591b23d7a016bf8d2a101cc65e7e

    DeviceType.ROBOT_CLEANER: (
        RUN_STATE_SENSOR_DESC[ThinQProperty.CURRENT_STATE],
        JOB_MODE_SENSOR_DESC[ThinQProperty.CURRENT_JOB_MODE],
        TIMER_SENSOR_DESC[TimerProperty.RUNNING],
    ),
    **DeviceType.STICK_CLEANER: (
        BATTERY_SENSOR_DESC[ThinQProperty.BATTERY_PERCENT],**
        JOB_MODE_SENSOR_DESC[ThinQPropertyEx.CURRENT_JOB_MODE_STICK_CLEANER],
        RUN_STATE_SENSOR_DESC[ThinQProperty.CURRENT_STATE],
    ),